File: packages/sdk/java/packages/admin/src/main/java/dev/edgebase/sdk/admin/D1Client.java
// EdgeBase Java SDK — D1Client for user-defined D1 databases.
package dev.edgebase.sdk.admin;

import dev.edgebase.sdk.core.*;

import java.util.*;

/**
 * Client for a user-defined D1 database.
 *
 * <pre>{@code
 * List<Object> rows = admin.d1("analytics").exec("SELECT * FROM events WHERE type = ?", List.of("click"));
 * }</pre>
 */
public class D1Client {
    private final HttpClient httpClient;
    private final String database;

    public D1Client(HttpClient httpClient, String database) {
        this.httpClient = httpClient;
        this.database = database;
    }

    /** Execute a SQL query without params. */
    public List<Object> exec(String query) {
        return exec(query, Collections.emptyList());
    }

    /**
     * Execute a SQL query with bind parameters.
     * All SQL is allowed (DDL included).
     */
    @SuppressWarnings("unchecked")
    public List<Object> exec(String query, List<Object> params) {
        Map<String, Object> body = new HashMap<>();
        body.put("query", query);
        if (!params.isEmpty())
            body.put("params", params);
        Object res = httpClient.post("/d1/" + database, body);
        if (res instanceof Map) {
            Object results = ((Map<String, Object>) res).get("results");
            return results instanceof List ? (List<Object>) results : Collections.emptyList();
        }
        return Collections.emptyList();
    }
}

File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/D1Client.kt
// D1Client — D1 database access for server-side use.
package dev.edgebase.sdk.admin

import dev.edgebase.sdk.core.*

/**
 * Client for a user-defined D1 database.
 *
 * ```kotlin
 * val rows = admin.d1("analytics").exec("SELECT * FROM events WHERE type = ?", listOf("click"))
 * ```
 */
class D1Client internal constructor(
    private val client: HttpClient,
    private val database: String
) {
    /**
     * Execute a SQL query. Use ? placeholders for bind parameters.
     * All SQL is allowed (DDL included).
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun exec(query: String, params: List<Any?> = emptyList()): List<Any?> {
        val body = mutableMapOf<String, Any?>("query" to query)
        if (params.isNotEmpty()) body["params"] = params
        val res = client.post("/d1/$database", body)
        val map = res as? Map<String, Any?> ?: return emptyList()
        return map["results"] as? List<Any?> ?: emptyList()
    }
}
